// react-rainbow-components/components/VerticalSectionOverflow/pageObject/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/**
 * VerticalSectionOverflow page object class.
 * @class
 */
var PageVerticalSectionOverflow =
/*#__PURE__*/
function () {
  /**
   * Create a new VerticalSectionOverflow page object.
   * @constructor
   * @param {string} rootElement - The selector of the VerticalSectionOverflow root element.
   */
  function PageVerticalSectionOverflow(rootElement) {
    (0, _classCallCheck2["default"])(this, PageVerticalSectionOverflow);
    this.rootElement = rootElement;
  }
  /**
   * Clicks the vertical section overflow button.
   * @method
   */


  (0, _createClass2["default"])(PageVerticalSectionOverflow, [{
    key: "click",
    value: function click() {
      $(this.rootElement).$('[data-id="vertical-overflow-button"]').click();
    }
    /**
     * Returns true when the overflow section is visible, false otherwise.
     * @method
     * @returns {bool}
     */

  }, {
    key: "isExpanded",
    value: function isExpanded() {
      return $(this.rootElement).$('[data-id="vertical-overflow"]').isDisplayed();
    }
    /**
     * Wait until the expand transition has finished.
     * @method
     */

  }, {
    key: "waitUntilExpand",
    value: function waitUntilExpand() {
      var _this = this;

      browser.waitUntil(function () {
        return _this.isExpanded();
      });
    }
    /**
     * Wait until the contract transition has finished.
     * @method
     */

  }, {
    key: "waitUntilCollapse",
    value: function waitUntilCollapse() {
      var _this2 = this;

      browser.waitUntil(function () {
        return !_this2.isExpanded();
      });
    }
    /**
     * Returns true when the vertical section overflow button has focus.
     * @method
     * @returns {bool}
     */

  }, {
    key: "hasFocusButton",
    value: function hasFocusButton() {
      return $(this.rootElement).$('[data-id="vertical-overflow-button"]').isFocused();
    }
  }]);
  return PageVerticalSectionOverflow;
}();

module.exports = PageVerticalSectionOverflow;
