// react-rainbow-components/components/Picklist/pageObject/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var PagePicklistOption = require('../../PicklistOption/pageObject');

var privateGetMenuBoundsRect = Symbol('privateGetMenuBoundsRect');
/**
 * Picklist page object class.
 * @class
 * @tutorial picklist
 */

var PagePicklist =
/*#__PURE__*/
function () {
  /**
   * Create a new PagePicklist page object.
   * @constructor
   * @param {string} rootElement - The selector of the PagePicklist root element.
   */
  function PagePicklist(rootElement) {
    (0, _classCallCheck2["default"])(this, PagePicklist);
    this.rootElement = rootElement;
  }
  /**
   * Clicks the input element.
   * @method
   */


  (0, _createClass2["default"])(PagePicklist, [{
    key: "clickInput",
    value: function clickInput() {
      $(this.rootElement).$('input[type="text"]').click();
    }
    /**
     * Focus the input element.
     * @method
     */

  }, {
    key: "focusInput",
    value: function focusInput() {
      this.clickInput();
      this.waitUntilOpen();
      this.clickInput();
    }
    /**
     * Returns true when the input element has focus.
     * @method
     * @returns {bool}
     */

  }, {
    key: "hasFocusInput",
    value: function hasFocusInput() {
      return $(this.rootElement).$('input[type="text"]').isFocused();
    }
    /**
     * It moves the pointer over the menu scroll up arrow
     * @method
     */

  }, {
    key: "hoverScrollUpArrow",
    value: function hoverScrollUpArrow() {
      return $(this.rootElement).$('[data-id=picklist-arrow-button-up]').moveTo();
    }
    /**
     * It moves the pointer out of the menu scroll up arrow
     * @method
     */

  }, {
    key: "mouseLeaveScrollUpArrow",
    value: function mouseLeaveScrollUpArrow() {
      return $(this.rootElement).$('input[type="text"]').moveTo();
    }
    /**
     * It moves the pointer over the menu scroll down arrow
     * @method
     */

  }, {
    key: "hoverScrollDownArrow",
    value: function hoverScrollDownArrow() {
      return $(this.rootElement).$('[data-id=picklist-arrow-button-down]').moveTo();
    }
    /**
     * It moves the pointer out of the menu scroll down arrow
     * @method
     */

  }, {
    key: "mouseLeaveScrollDownArrow",
    value: function mouseLeaveScrollDownArrow() {
      return $(this.rootElement).$('input[type="text"]').moveTo();
    }
    /**
     * Returns the label of the selected PicklistOption
     * @method
     * @returns {string}
     */

  }, {
    key: "getSelectedOptionLabel",
    value: function getSelectedOptionLabel() {
      return $(this.rootElement).$('input[type="text"]').getValue();
    }
    /**
     * Get the number of registered options.
     * @method
     * @returns {number}
     */

  }, {
    key: "getOptionsLength",
    value: function getOptionsLength() {
      return $(this.rootElement).$$('li[data-selected="false"]').length;
    }
    /**
     * Returns a new PicklistOption page object of the element in item position.
     * @method
     * @param {number} optionIndex - The base 0 index of the PicklistOption.
     */

  }, {
    key: "getOption",
    value: function getOption(optionIndex) {
      var activeOptions = $(this.rootElement).$$('li[data-selected="false"]');
      var option = activeOptions[optionIndex];

      if (option && !option.error) {
        return new PagePicklistOption(option, this[privateGetMenuBoundsRect]());
      }

      return null;
    }
    /**
     * Returns true when the options menu is open, false otherwise.
     * @method
     * @returns {bool}
     */

  }, {
    key: "isMenuOpen",
    value: function isMenuOpen() {
      return $(this.rootElement).$('[role="listbox"]').isDisplayed();
    }
    /**
     * Returns the boundaries of Picklist dropdown menu.
     * @method
     * @returns {object}
     */

  }, {
    key: privateGetMenuBoundsRect,
    value: function value() {
      var menu = $(this.rootElement).$('[role="listbox"]');

      var _menu$getLocation = menu.getLocation(),
          x = _menu$getLocation.x,
          y = _menu$getLocation.y;

      var _menu$getSize = menu.getSize(),
          width = _menu$getSize.width,
          height = _menu$getSize.height;

      return {
        left: x,
        top: y,
        right: x + width,
        bottom: y + height
      };
    }
    /**
     * Wait until the options menu is open.
     * @method
     */

  }, {
    key: "waitUntilOpen",
    value: function waitUntilOpen() {
      var _this = this;

      browser.waitUntil(function () {
        return _this.isMenuOpen();
      });
    }
  }]);
  return PagePicklist;
}();

module.exports = PagePicklist;
