// react-rainbow-components/components/TimelineMarker/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = TimelineMarker;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _calendarIcon = _interopRequireDefault(require("./calendarIcon"));

var _li = _interopRequireDefault(require("./styled/li"));

var _columnLeft = _interopRequireDefault(require("./styled/columnLeft"));

var _iconContainer = _interopRequireDefault(require("./styled/iconContainer"));

var _contentContainer = _interopRequireDefault(require("./styled/contentContainer"));

var _header = _interopRequireDefault(require("./styled/header"));

var _label = _interopRequireDefault(require("./styled/label"));

var _datetime = _interopRequireDefault(require("./styled/datetime"));

var _description = _interopRequireDefault(require("./styled/description"));

var _body = _interopRequireDefault(require("./styled/body"));

/**
 * The TimelineMarker displays one event of an item timeline.
 * @category Layout
 */
function TimelineMarker(props) {
  var icon = props.icon,
      label = props.label,
      description = props.description,
      datetime = props.datetime,
      children = props.children,
      className = props.className,
      style = props.style;
  return _react["default"].createElement(_li["default"], {
    className: className,
    style: style
  }, _react["default"].createElement(_columnLeft["default"], null, _react["default"].createElement(_iconContainer["default"], null, icon)), _react["default"].createElement(_contentContainer["default"], null, _react["default"].createElement("div", null, _react["default"].createElement(_header["default"], null, _react["default"].createElement(_label["default"], null, label), _react["default"].createElement(_datetime["default"], null, datetime)), _react["default"].createElement(_description["default"], null, description)), _react["default"].createElement(_body["default"], null, children)));
}

TimelineMarker.propTypes = {
  /** The text to be displayed as the TimelineMarker's label. */
  label: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** The text to be displayed as the TimelineMarker's description. */
  description: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** The text with the formatted datetime of the event. */
  datetime: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** The icon to show at the left of the TimelineMarker.
   * If not passed by default a calendar icon will be showed.  */
  icon: _propTypes["default"].node,

  /**
   * This prop that should not be visible in the documentation.
   * @ignore
   */
  children: _propTypes["default"].oneOfType([_propTypes["default"].arrayOf(_propTypes["default"].node), _propTypes["default"].object]),

  /** A CSS class for the outer element, in addition to the component's base classes. */
  className: _propTypes["default"].string,

  /** An object with custom style applied to the outer element. */
  style: _propTypes["default"].object
};
TimelineMarker.defaultProps = {
  label: null,
  description: null,
  datetime: null,
  icon: _react["default"].createElement(_calendarIcon["default"], null),
  children: null,
  className: undefined,
  style: undefined
};
