// react-rainbow-components/components/ImportRecordsFlow/stepTwo/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = StepTwo;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _uploadFileButton = _interopRequireDefault(require("./uploadFileButton"));

var _preview = _interopRequireDefault(require("./preview"));

var _container = _interopRequireDefault(require("./styled/container"));

var _text = _interopRequireDefault(require("./styled/text"));

function preventDefault(event) {
  event.stopPropagation();
  event.preventDefault();
}

function StepTwo(props) {
  var onProcessFile = props.onProcessFile,
      isLoading = props.isLoading,
      hasFileSelected = props.hasFileSelected,
      columns = props.columns,
      data = props.data,
      onRemoveFile = props.onRemoveFile,
      fileName = props.fileName,
      fileType = props.fileType;

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      isDragOver = _useState2[0],
      setIsDragOver = _useState2[1];

  var handleChange = function handleChange(event) {
    onProcessFile(event.target.files[0]);
  };

  var handleDragLeave = function handleDragLeave() {
    setIsDragOver(false);
  };

  var handleDragOver = function handleDragOver(event) {
    preventDefault(event);
    setIsDragOver(true);
  };

  var handleDrop = function handleDrop(event) {
    preventDefault(event);
    setIsDragOver(false);
    var files = event.dataTransfer.files;

    if (files.length === 1 && files[0].type === 'text/csv') {
      onProcessFile(files[0]);
    }
  };

  if (hasFileSelected) {
    return _react["default"].createElement(_preview["default"], {
      fileName: fileName,
      fileType: fileType,
      columns: columns,
      data: data,
      isLoading: isLoading,
      onRemoveFile: onRemoveFile
    });
  }

  return _react["default"].createElement(_container["default"], {
    isDragOver: isDragOver,
    onDragEnter: preventDefault,
    onDragLeave: handleDragLeave,
    onDragOver: handleDragOver,
    onDrop: handleDrop,
    draggable: true
  }, _react["default"].createElement(_uploadFileButton["default"], {
    onChange: handleChange
  }), _react["default"].createElement(_text["default"], null, "Find the file in your desktop."));
}

StepTwo.propTypes = {
  onProcessFile: _propTypes["default"].func,
  isLoading: _propTypes["default"].bool,
  hasFileSelected: _propTypes["default"].bool,
  columns: _propTypes["default"].array,
  data: _propTypes["default"].array,
  onRemoveFile: _propTypes["default"].func,
  fileName: _propTypes["default"].string,
  fileType: _propTypes["default"].string
};
StepTwo.defaultProps = {
  onProcessFile: function onProcessFile() {},
  isLoading: false,
  hasFileSelected: false,
  columns: [],
  data: [],
  onRemoveFile: function onRemoveFile() {},
  fileName: '',
  fileType: ''
};
