// react-rainbow-components/components/MenuItem/pageObject/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/**
 * MenuItem page object class.
 * @class
 */
var PageMenuItem =
/*#__PURE__*/
function () {
  /**
   * Create a new MenuItem page object.
   * @constructor
   * @param {string} rootElement - The selector of the MenuItem root element.
   */
  function PageMenuItem(rootElement) {
    (0, _classCallCheck2["default"])(this, PageMenuItem);
    this.rootElement = rootElement;
  }
  /**
   * Clicks the menu item.
   * @method
   */


  (0, _createClass2["default"])(PageMenuItem, [{
    key: "click",
    value: function click() {
      $(this.rootElement).click();
    }
    /**
     * Returns true when the menu item has focus.
     * @method
     * @returns {bool}
     */

  }, {
    key: "hasFocus",
    value: function hasFocus() {
      return $(this.rootElement).isFocused();
    }
    /**
     * It moves the pointer over the menu item.
     * @method
     */

  }, {
    key: "hover",
    value: function hover() {
      $(this.rootElement).moveTo();
    }
    /**
     * Returns the label text of the menu item.
     * @method
     * @returns {string}
     */

  }, {
    key: "getLabelText",
    value: function getLabelText() {
      return $(this.rootElement).getText();
    }
  }]);
  return PageMenuItem;
}();

module.exports = PageMenuItem;
