// react-rainbow-components/components/VerticalNavigation/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = VerticalNavigation;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _context = require("./context");

var _nav = _interopRequireDefault(require("./styled/nav"));

/**
 * Navigation represents a list of links that either take the user to another page
 * or parts of the page the user is in.
 * @category Layout
 */
function VerticalNavigation(props) {
  var id = props.id,
      ariaLabel = props.ariaLabel,
      style = props.style,
      selectedItem = props.selectedItem,
      onSelect = props.onSelect,
      compact = props.compact,
      shaded = props.shaded,
      className = props.className,
      children = props.children;
  var context = {
    selectedItem: selectedItem,
    onSelect: onSelect
  };
  return _react["default"].createElement(_nav["default"], {
    id: id,
    className: className,
    style: style,
    "aria-label": ariaLabel,
    compact: compact,
    shaded: shaded
  }, _react["default"].createElement(_context.Provider, {
    value: context
  }, children));
}

VerticalNavigation.propTypes = {
  /** Name of the nagivation item to make active. */
  selectedItem: _propTypes["default"].string,

  /** Action fired when an item is selected.
   * The event params include the `name` of the selected item. */
  onSelect: _propTypes["default"].func,

  /** Specify true to reduce spacing between navigation items. This value defaults to false. */
  compact: _propTypes["default"].bool,

  /** Specify true when the vertical navigation is sitting on top of a shaded background.
   * This value defaults to false. */
  shaded: _propTypes["default"].bool,

  /** The aria label attribute for the navigation component. */
  ariaLabel: _propTypes["default"].string,

  /** A CSS class for the outer element, in addition to the component's base classes. */
  className: _propTypes["default"].string,

  /** An object with custom style applied for the outer element. */
  style: _propTypes["default"].object,

  /**
   * This prop that should not be visible in the documentation.
   * @ignore
   */
  children: _propTypes["default"].node,

  /** The id of the outer element. */
  id: _propTypes["default"].string
};
VerticalNavigation.defaultProps = {
  selectedItem: null,
  onSelect: function onSelect() {},
  compact: false,
  shaded: false,
  ariaLabel: undefined,
  className: undefined,
  style: undefined,
  children: null,
  id: undefined
};
