// react-rainbow-components/components/DateTimePicker/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _withReduxForm = _interopRequireDefault(require("../../libs/hocs/withReduxForm"));

var _icon = _interopRequireDefault(require("./icon"));

var _pickerModal = _interopRequireDefault(require("./pickerModal"));

var _formatDateTime = _interopRequireDefault(require("./helpers/formatDateTime"));

var _constants = require("../../libs/constants");

var _container = _interopRequireDefault(require("./styled/container"));

var _input = _interopRequireDefault(require("./styled/input"));

var _context = require("../Application/context");

var _utils = require("../../libs/utils");

/* eslint-disable react/prop-types */

/**
 * A DateTimePicker is used to select a day and a time.
 * @category Form
 */
var DateTimePicker = _react["default"].forwardRef(function (props, ref) {
  var placeholder = props.placeholder,
      hideLabel = props.hideLabel,
      required = props.required,
      name = props.name,
      label = props.label,
      error = props.error,
      readOnly = props.readOnly,
      disabled = props.disabled,
      tabIndex = props.tabIndex,
      onClick = props.onClick,
      onChange = props.onChange,
      onFocus = props.onFocus,
      onBlur = props.onBlur,
      id = props.id,
      className = props.className,
      style = props.style,
      value = props.value,
      minDate = props.minDate,
      maxDate = props.maxDate,
      formatStyle = props.formatStyle,
      okLabel = props.okLabel,
      cancelLabel = props.cancelLabel,
      isCentered = props.isCentered,
      bottomHelpText = props.bottomHelpText,
      hour24 = props.hour24,
      localeProp = props.locale;
  var inputRef = (0, _react.useRef)();
  var modalId = id && "".concat(id, "_modal");
  (0, _react.useImperativeHandle)(ref, function () {
    return {
      focus: function focus() {
        inputRef.current.focus();
      },
      click: function click() {
        inputRef.current.click();
      },
      blur: function blur() {
        inputRef.current.blur();
      }
    };
  });
  var context = (0, _react.useContext)(_context.AppContext);
  var locale = (0, _utils.getLocale)(context, localeProp);

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      isOpen = _useState2[0],
      setIsOpen = _useState2[1];

  var _useState3 = (0, _react.useState)((0, _formatDateTime["default"])(value, formatStyle, locale, hour24)),
      _useState4 = (0, _slicedToArray2["default"])(_useState3, 2),
      formattedDatetime = _useState4[0],
      setFormattedDatetime = _useState4[1];

  (0, _react.useEffect)(function () {
    setFormattedDatetime((0, _formatDateTime["default"])(value, formatStyle, locale, hour24));
  }, [value, formatStyle, locale, hour24]);

  var openModal = function openModal(event) {
    if (!readOnly) {
      setIsOpen(true);
      onClick(event);
    }
  };

  var closeModal = function closeModal() {
    setIsOpen(false);
  };

  var handleFocus = function handleFocus() {
    onFocus(value);
  };

  var handleBlur = function handleBlur() {
    onBlur(value);
  };

  var handleKeyDown = function handleKeyDown(_ref) {
    var keyCode = _ref.keyCode;
    var shouldOpenModal = (keyCode === _constants.ENTER_KEY || keyCode === _constants.SPACE_KEY) && !readOnly;

    if (shouldOpenModal) {
      setIsOpen(true);
    }
  };

  var handleChange = function handleChange() {
    closeModal();
    onChange.apply(void 0, arguments);
  };

  return _react["default"].createElement(_container["default"], {
    id: id,
    className: className,
    style: style
  }, _react["default"].createElement(_input["default"], {
    ref: inputRef,
    label: label,
    placeholder: placeholder,
    icon: _react["default"].createElement(_icon["default"], null),
    iconPosition: "right",
    required: required,
    value: formattedDatetime,
    onKeyDown: handleKeyDown,
    onClick: openModal,
    onFocus: handleFocus,
    onBlur: handleBlur,
    hideLabel: hideLabel,
    name: name,
    bottomHelpText: bottomHelpText,
    isCentered: isCentered,
    error: error,
    readOnly: readOnly,
    disabled: disabled,
    tabIndex: tabIndex,
    autoComplete: "off"
  }), _react["default"].createElement(_pickerModal["default"], {
    id: modalId,
    isOpen: isOpen,
    title: formattedDatetime,
    onRequestClose: closeModal,
    formatStyle: formatStyle,
    value: value,
    onChange: handleChange,
    minDate: minDate,
    maxDate: maxDate,
    okLabel: okLabel,
    cancelLabel: cancelLabel,
    locale: locale,
    hour24: hour24
  }));
});

DateTimePicker.propTypes = {
  /** Sets the date for the DateTimePicker programmatically. */
  value: _propTypes["default"].oneOfType([_propTypes["default"].instanceOf(Date), _propTypes["default"].string]),

  /** The beginning of a range of valid dates. The range includes the startDate.
   * The default value is current date - 100 years. */
  minDate: _propTypes["default"].instanceOf(Date),

  /** The ending of a range of valid dates. The range includes the endDate.
   * The default value is current date + 100 years. */
  maxDate: _propTypes["default"].instanceOf(Date),

  /**  The date time format style to display in the input field.
   * Valid values are small, medium, and large. */
  formatStyle: _propTypes["default"].oneOf(['small', 'medium', 'large']),

  /** The name of the DateTimePicker. */
  name: _propTypes["default"].string,

  /** Text label for the DateTimePicker. */
  label: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** A boolean to hide the DateTimePicker label. */
  hideLabel: _propTypes["default"].bool,

  /** Text that is displayed when the DateTimePicker is empty,
   * to prompt the user for a valid entry. */
  placeholder: _propTypes["default"].string,

  /** Specifies that the DateTimePicker field must be filled out before submitting the form.
   * This value defaults to false. */
  required: _propTypes["default"].bool,

  /** Shows the help message below the DateTimePicker. */
  bottomHelpText: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** Specifies that the DateTimePicker text will be centered. This value defaults to false. */
  isCentered: _propTypes["default"].bool,

  /** Specifies that the DateTimePicker must be filled out before submitting the form. */
  error: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** Specifies that the DateTimePicker is read-only. This value defaults to false. */
  readOnly: _propTypes["default"].bool,

  /** Specifies that the DateTimePicker element should be disabled.
   * This value defaults to false. */
  disabled: _propTypes["default"].bool,

  /** Specifies the tab order of an element (when the tab button is used for navigating). */
  tabIndex: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),

  /** The action triggered when a value attribute changes. */
  onChange: _propTypes["default"].func,

  /** The action triggered when the element is clicked. */
  onClick: _propTypes["default"].func,

  /** The action triggered when the element receives focus. */
  onFocus: _propTypes["default"].func,

  /** The action triggered when the element releases focus. */
  onBlur: _propTypes["default"].func,

  /** The id of the outer element. */
  id: _propTypes["default"].string,

  /** A CSS class for the outer element, in addition to the component's base classes. */
  className: _propTypes["default"].string,

  /** An object with custom style applied to the outer element. */
  style: _propTypes["default"].object,

  /** Text label for the OK button in the modal dialog. */
  okLabel: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** Text label for the CANCEL button in the modal dialog. */
  cancelLabel: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),

  /** The DateTimePicker locale. Defaults to browser's language. */
  locale: _propTypes["default"].string,

  /** Specifies that the DateTimePicker will be in a 24hr format. This value defaults to false. */
  hour24: _propTypes["default"].bool
};
DateTimePicker.defaultProps = {
  placeholder: undefined,
  label: undefined,
  hideLabel: false,
  required: false,
  name: undefined,
  error: null,
  readOnly: false,
  disabled: false,
  tabIndex: undefined,
  onClick: function onClick() {},
  onChange: function onChange() {},
  onFocus: function onFocus() {},
  onBlur: function onBlur() {},
  id: undefined,
  className: undefined,
  style: undefined,
  value: undefined,
  minDate: undefined,
  maxDate: undefined,
  formatStyle: 'medium',
  okLabel: 'Ok',
  cancelLabel: 'Cancel',
  bottomHelpText: '',
  isCentered: false,
  locale: undefined,
  hour24: false
};

var _default = (0, _withReduxForm["default"])(DateTimePicker);

exports["default"] = _default;
