// react-rainbow-components/components/Textarea/pageObject/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/**
 * Textarea page object class.
 * @class
 */
var PageTextarea =
/*#__PURE__*/
function () {
  /**
   * Create a new Textarea page object.
   * @constructor
   * @param {string} rootElement - The selector of the Textarea root element.
   */
  function PageTextarea(rootElement) {
    (0, _classCallCheck2["default"])(this, PageTextarea);
    this.rootElement = rootElement;
  }
  /**
   * Clicks the textarea element.
   * @method
   */


  (0, _createClass2["default"])(PageTextarea, [{
    key: "click",
    value: function click() {
      $(this.rootElement).$('textarea').click();
    }
    /**
     * Clicks the label element.
     * @method
     */

  }, {
    key: "clickLabel",
    value: function clickLabel() {
      $(this.rootElement).$('label').click();
    }
    /**
     * Returns true when the textarea element has focus.
     * @method
     * @returns {bool}
     */

  }, {
    key: "hasFocusTextarea",
    value: function hasFocusTextarea() {
      return $(this.rootElement).$('textarea').isFocused();
    }
    /**
     * It set a value in the textarea.
     * @method
     * @param {string} values - The value to set in the textarea.
     */

  }, {
    key: "setValue",
    value: function setValue(value) {
      $(this.rootElement).$('textarea').setValue(value);
    }
    /**
     * It get the value of the textarea.
     * @method
     * @returns {string}
     */

  }, {
    key: "getValue",
    value: function getValue() {
      return $(this.rootElement).$('textarea').getValue();
    }
  }]);
  return PageTextarea;
}();

module.exports = PageTextarea;
