// react-rainbow-components/components/Table/body/actionsCell.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ActionsCell;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ButtonMenu = _interopRequireDefault(require("../../ButtonMenu"));

var _more = _interopRequireDefault(require("./icons/more"));

var _cellContainer = _interopRequireDefault(require("./styled/cellContainer"));

var _actionCell = _interopRequireDefault(require("./styled/actionCell"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function MenuItems(_ref) {
  var children = _ref.children,
      rowData = _ref.rowData;
  return _react["default"].Children.map(children, function (child) {
    if ((0, _react.isValidElement)(child)) {
      return _objectSpread({}, child, {
        props: _objectSpread({}, child.props, {
          onClick: function onClick(event) {
            return child.props.onClick(event, rowData);
          }
        })
      });
    }

    return null;
  }, null);
}

function ActionsCell(props) {
  var columnChildren = props.columnChildren,
      rowsLength = props.rowsLength,
      rowIndex = props.rowIndex,
      rowData = props.rowData;

  if (columnChildren) {
    var getMenuAlignment = function getMenuAlignment() {
      var menuItemsLength = columnChildren.length || 1;
      var rowLengthToBottom = rowsLength - (rowIndex + 1);
      var alignMenuAtBottom;

      if (rowsLength / 2 > menuItemsLength) {
        alignMenuAtBottom = rowLengthToBottom <= menuItemsLength;
      } else {
        alignMenuAtBottom = rowIndex > menuItemsLength;
      }

      if (alignMenuAtBottom) {
        return 'bottom-right';
      }

      return 'right';
    };

    return _react["default"].createElement(_cellContainer["default"], {
      role: "gridcell",
      tabIndex: -1
    }, _react["default"].createElement(_actionCell["default"], null, _react["default"].createElement(_ButtonMenu["default"], {
      tabIndex: -1,
      icon: _react["default"].createElement(_more["default"], null),
      menuAlignment: getMenuAlignment(),
      buttonSize: "small"
    }, _react["default"].createElement(MenuItems, {
      rowData: rowData
    }, columnChildren))));
  }

  return null;
}

ActionsCell.propTypes = {
  columnChildren: _propTypes["default"].oneOfType([_propTypes["default"].arrayOf(_propTypes["default"].node), _propTypes["default"].object]),
  rowsLength: _propTypes["default"].number,
  rowIndex: _propTypes["default"].number,
  rowData: _propTypes["default"].object
};
ActionsCell.defaultProps = {
  columnChildren: undefined,
  rowsLength: undefined,
  rowIndex: undefined,
  rowData: {}
};
