// react-rainbow-components/components/Input/inputRadio/radio.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _utils = require("../../../libs/utils");

var _checkboxRadioLabel = _interopRequireDefault(require("../checkboxRadioLabel"));

var _container = _interopRequireDefault(require("./styled/container"));

var _radio = _interopRequireDefault(require("./styled/radio"));

var Radio =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(Radio, _Component);

  function Radio(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Radio);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Radio).call(this, props));
    _this.inputId = (0, _utils.uniqueId)('input-radio');
    _this.inputRef = _react["default"].createRef();
    return _this;
  }
  /**
   * Sets focus on the element.
   * @public
   */


  (0, _createClass2["default"])(Radio, [{
    key: "focus",
    value: function focus() {
      this.inputRef.current.focus();
    }
    /**
     * Sets click on the element.
     * @public
     */

  }, {
    key: "click",
    value: function click() {
      this.inputRef.current.click();
    }
    /**
     * Sets blur on the element.
     * @public
     */

  }, {
    key: "blur",
    value: function blur() {
      this.inputRef.current.blur();
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          name = _this$props.name,
          value = _this$props.value,
          onChange = _this$props.onChange,
          tabIndex = _this$props.tabIndex,
          onFocus = _this$props.onFocus,
          onBlur = _this$props.onBlur,
          onClick = _this$props.onClick,
          onKeyDown = _this$props.onKeyDown,
          disabled = _this$props.disabled,
          checked = _this$props.checked,
          ariaLabelledBy = _this$props.ariaLabelledBy,
          ariaDescribedBy = _this$props.ariaDescribedBy,
          error = _this$props.error,
          label = _this$props.label,
          hideLabel = _this$props.hideLabel;
      return _react["default"].createElement(_container["default"], {
        "data-id": "input-radio_container"
      }, _react["default"].createElement(_radio["default"], {
        as: "input",
        id: this.inputId,
        name: name,
        type: "radio",
        value: value,
        onChange: onChange,
        tabIndex: tabIndex,
        onFocus: onFocus,
        onBlur: onBlur,
        onClick: onClick,
        onKeyDown: onKeyDown,
        disabled: disabled,
        checked: checked,
        "aria-labelledby": ariaLabelledBy,
        "aria-describedby": ariaDescribedBy,
        ref: this.inputRef,
        error: error
      }), _react["default"].createElement(_checkboxRadioLabel["default"], {
        label: label,
        hideLabel: hideLabel,
        disabled: disabled,
        inputId: this.inputId,
        id: ariaLabelledBy
      }));
    }
  }]);
  return Radio;
}(_react.Component);

exports["default"] = Radio;
Radio.propTypes = {
  value: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].bool]),
  label: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),
  name: _propTypes["default"].string,
  error: _propTypes["default"].oneOfType([_propTypes["default"].string, _propTypes["default"].node]),
  disabled: _propTypes["default"].bool,
  onChange: _propTypes["default"].func,
  tabIndex: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  onClick: _propTypes["default"].func,
  onFocus: _propTypes["default"].func,
  onBlur: _propTypes["default"].func,
  onKeyDown: _propTypes["default"].func,
  checked: _propTypes["default"].bool,
  hideLabel: _propTypes["default"].bool,
  ariaLabelledBy: _propTypes["default"].string,
  ariaDescribedBy: _propTypes["default"].string
};
Radio.defaultProps = {
  value: undefined,
  label: undefined,
  name: undefined,
  error: null,
  disabled: false,
  onChange: function onChange() {},
  tabIndex: undefined,
  onClick: function onClick() {},
  onFocus: function onFocus() {},
  onBlur: function onBlur() {},
  onKeyDown: function onKeyDown() {},
  checked: undefined,
  hideLabel: false,
  ariaLabelledBy: undefined,
  ariaDescribedBy: undefined
};
