// react-rainbow-components/components/AvatarMenu/avatarButton.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _Avatar = _interopRequireDefault(require("./../Avatar"));

var _button = _interopRequireDefault(require("./styled/button"));

var AvatarButton =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(AvatarButton, _Component);

  function AvatarButton(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, AvatarButton);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(AvatarButton).call(this, props));
    _this.avatarButtonRef = _react["default"].createRef();
    return _this;
  }
  /**
   * Sets focus on the element.
   * @public
   */


  (0, _createClass2["default"])(AvatarButton, [{
    key: "focus",
    value: function focus() {
      this.avatarButtonRef.current.focus();
    }
    /**
     * Sets click on the element.
     * @public
     */

  }, {
    key: "click",
    value: function click() {
      this.avatarButtonRef.current.click();
    }
    /**
     * Sets blur on the element.
     * @public
     */

  }, {
    key: "blur",
    value: function blur() {
      this.avatarButtonRef.current.blur();
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          title = _this$props.title,
          tabIndex = _this$props.tabIndex,
          onClick = _this$props.onClick,
          onFocus = _this$props.onFocus,
          onBlur = _this$props.onBlur,
          disabled = _this$props.disabled,
          assistiveText = _this$props.assistiveText,
          ariaHaspopup = _this$props.ariaHaspopup,
          src = _this$props.src,
          initials = _this$props.initials,
          icon = _this$props.icon,
          avatarSize = _this$props.avatarSize,
          initialsVariant = _this$props.initialsVariant;
      return _react["default"].createElement(_button["default"], {
        "data-id": "avatar-menu-button",
        tabIndex: tabIndex,
        onFocus: onFocus,
        onBlur: onBlur,
        disabled: disabled,
        onClick: onClick,
        title: title,
        "aria-haspopup": ariaHaspopup,
        ref: this.avatarButtonRef
      }, _react["default"].createElement(_Avatar["default"], {
        src: src,
        icon: icon,
        initials: initials,
        size: avatarSize,
        initialsVariant: initialsVariant,
        title: title,
        assistiveText: assistiveText,
        ariaHaspopup: true,
        onFocus: onFocus,
        onBlur: onBlur
      }));
    }
  }]);
  return AvatarButton;
}(_react.Component);

exports["default"] = AvatarButton;
AvatarButton.propTypes = {
  title: _propTypes["default"].string,
  tabIndex: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
  onClick: _propTypes["default"].func,
  onFocus: _propTypes["default"].func,
  onBlur: _propTypes["default"].func,
  assistiveText: _propTypes["default"].string,
  ariaHaspopup: _propTypes["default"].bool,
  disabled: _propTypes["default"].bool,
  src: _propTypes["default"].string,
  initials: _propTypes["default"].string,
  icon: _propTypes["default"].node,
  avatarSize: _propTypes["default"].oneOf(['x-small', 'small', 'medium', 'large']),
  initialsVariant: _propTypes["default"].oneOf(['default', 'inverse'])
};
AvatarButton.defaultProps = {
  title: undefined,
  disabled: false,
  tabIndex: undefined,
  onClick: function onClick() {},
  onFocus: function onFocus() {},
  onBlur: function onBlur() {},
  assistiveText: undefined,
  ariaHaspopup: false,
  src: undefined,
  initials: undefined,
  icon: null,
  avatarSize: 'medium',
  initialsVariant: 'default'
};
