// react-rainbow-components/components/ButtonMenu/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ButtonMenu;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ButtonIcon = _interopRequireDefault(require("./../ButtonIcon"));

var _PrimitiveMenu = _interopRequireDefault(require("../PrimitiveMenu"));

/**
 * A Button Menu offers a list of actions or functions that a user can access.
 */
function ButtonMenu(props) {
  var icon = props.icon,
      buttonSize = props.buttonSize,
      title = props.title,
      assistiveText = props.assistiveText,
      buttonVariant = props.buttonVariant,
      buttonShaded = props.buttonShaded,
      disabled = props.disabled,
      tabIndex = props.tabIndex,
      onClick = props.onClick,
      onFocus = props.onFocus,
      onBlur = props.onBlur,
      children = props.children,
      rest = (0, _objectWithoutProperties2["default"])(props, ["icon", "buttonSize", "title", "assistiveText", "buttonVariant", "buttonShaded", "disabled", "tabIndex", "onClick", "onFocus", "onBlur", "children"]);
  return _react["default"].createElement(_PrimitiveMenu["default"], (0, _extends2["default"])({}, rest, {
    icon: icon,
    size: buttonSize,
    assistiveText: assistiveText,
    disabled: disabled,
    tabIndex: tabIndex,
    variant: buttonVariant,
    shaded: buttonShaded,
    title: title,
    onClick: onClick,
    onFocus: onFocus,
    onBlur: onBlur,
    trigger: _ButtonIcon["default"]
  }), children);
}

ButtonMenu.propTypes = {
  /** The icon to show if it is passed.
   * It must be a svg icon or a font icon. It is a required value. */
  icon: _propTypes["default"].node,

  /** The content of the ButtonMenu. Used to render the menuItem elements
   * when the ButtonMenu is open. */
  children: _propTypes["default"].oneOfType([_propTypes["default"].arrayOf(_propTypes["default"].node), _propTypes["default"].object]),

  /** The variant changes the look of the button. Accepted variants include base,
   * brand, success, border, border-filled, bare-inverse, and inverse.
   * This value defaults to border-filled. */
  buttonVariant: _propTypes["default"].oneOf(['base', 'brand', 'success', 'border', 'border-filled', 'border-inverse', 'inverse']),

  /** The size of the button. Options include xx-small, x-small, medium, or large.
   * This value defaults to medium. */
  buttonSize: _propTypes["default"].oneOf(['xx-small', 'x-small', 'small', 'medium', 'large']),

  /** Specify true when the button has a shadow around it.
   * This value defaults to false.
   * Only neutral, brand, destructive and success variant can be shaded. */
  buttonShaded: _propTypes["default"].bool,

  /** The size of the menu. Options include xx-small, x-small, medium, or large.
   * This value defaults to small. */
  menuSize: _propTypes["default"].oneOf(['xx-small', 'x-small', 'small', 'medium', 'large']),

  /** Determines the alignment of the menu relative to the button.
   * Available options are: left, center, right, bottom, bottom-left, bottom-right.
   * This value defaults to left. */
  menuAlignment: _propTypes["default"].oneOf(['left', 'right', 'bottom', 'center', 'bottom-right', 'bottom-left']),

  /** If true, the menu is disabled. Disabling the menu prevents users from opening it.
   * This value defaults to false. */
  disabled: _propTypes["default"].bool,

  /** If is set to true, then is showed a loading symbol. */
  isLoading: _propTypes["default"].bool,

  /** Displays tooltip text when the mouse moves over the element. */
  title: _propTypes["default"].string,

  /** A description for assistive sreen readers. */
  assistiveText: _propTypes["default"].string,

  /** Specifies the tab order of an element (when the tab button is used for navigating). */
  tabIndex: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),

  /** The action triggered when the element is clicked. */
  onClick: _propTypes["default"].func,

  /** The action triggered when the element receives focus. */
  onFocus: _propTypes["default"].func,

  /** The action triggered when the element releases focus. */
  onBlur: _propTypes["default"].func,

  /** A CSS class for the outer element, in addition to the component's base classes. */
  className: _propTypes["default"].string,

  /** An object with custom style applied to the outer element. */
  style: _propTypes["default"].object,

  /** The id of the outer element. */
  id: _propTypes["default"].string
};
ButtonMenu.defaultProps = {
  icon: null,
  children: null,
  buttonVariant: 'border-filled',
  buttonSize: 'medium',
  buttonShaded: false,
  menuSize: 'xx-small',
  menuAlignment: 'left',
  disabled: false,
  isLoading: false,
  tabIndex: undefined,
  title: undefined,
  assistiveText: undefined,
  onClick: function onClick() {},
  onFocus: function onFocus() {},
  onBlur: function onBlur() {},
  className: undefined,
  style: undefined,
  id: undefined
};
