// react-rainbow-components/libs/hocs/withDebounce/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = withDebounce;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _debounce = _interopRequireDefault(require("../../debounce"));

/* eslint-disable react/prop-types */
function withDebounce(WrappedComponent) {
  return (
    /*#__PURE__*/
    function (_Component) {
      (0, _inherits2["default"])(_class, _Component);

      function _class(props) {
        var _this;

        (0, _classCallCheck2["default"])(this, _class);
        _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(_class).call(this, props));
        _this.fieldRef = _react["default"].createRef();
        _this.handleChange = _this.handleChange.bind((0, _assertThisInitialized2["default"])(_this));
        _this.debouncedChange = (0, _debounce["default"])(_this.debouncedChange, props.debounceTimeout);
        return _this;
      }

      (0, _createClass2["default"])(_class, [{
        key: "handleChange",
        value: function handleChange(event) {
          event.persist();
          this.debouncedChange(event);
        }
      }, {
        key: "debouncedChange",
        value: function debouncedChange(event) {
          var onChange = this.props.onChange;
          onChange(event);
        }
      }, {
        key: "focus",
        value: function focus() {
          this.fieldRef.current.focus();
        }
      }, {
        key: "click",
        value: function click() {
          this.fieldRef.current.click();
        }
      }, {
        key: "blur",
        value: function blur() {
          this.fieldRef.current.blur();
        }
      }, {
        key: "render",
        value: function render() {
          var _this$props = this.props,
              onChange = _this$props.onChange,
              rest = (0, _objectWithoutProperties2["default"])(_this$props, ["onChange"]);
          return _react["default"].createElement(WrappedComponent, (0, _extends2["default"])({}, rest, {
            onChange: this.handleChange,
            ref: this.fieldRef
          }));
        }
      }]);
      return _class;
    }(_react.Component)
  );
}
