// react-rainbow-components/components/Rating/star.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _react = _interopRequireWildcard(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _AssistiveText = _interopRequireDefault(require("../AssistiveText"));

var _utils = require("../../libs/utils");

var _starFill = _interopRequireDefault(require("./starFill"));

var _starBordered = _interopRequireDefault(require("./starBordered"));

var _starContainer = _interopRequireDefault(require("./styled/starContainer"));

var _starInput = _interopRequireDefault(require("./styled/starInput"));

var _starHalf = _interopRequireDefault(require("./starHalf"));

var Star =
/*#__PURE__*/
function (_Component) {
  (0, _inherits2["default"])(Star, _Component);

  function Star(props) {
    var _this;

    (0, _classCallCheck2["default"])(this, Star);
    _this = (0, _possibleConstructorReturn2["default"])(this, (0, _getPrototypeOf2["default"])(Star).call(this, props));
    _this.starId = (0, _utils.uniqueId)('star');
    _this.handleChange = _this.handleChange.bind((0, _assertThisInitialized2["default"])(_this));
    return _this;
  }

  (0, _createClass2["default"])(Star, [{
    key: "getAssitiveText",
    value: function getAssitiveText() {
      var value = this.props.value;

      if (value === 1) {
        return "".concat(value, " Star");
      }

      return "".concat(value, " Stars");
    }
  }, {
    key: "handleChange",
    value: function handleChange(event) {
      var _this$props = this.props,
          readOnly = _this$props.readOnly,
          onChange = _this$props.onChange;
      if (readOnly) return null;
      return onChange(event);
    }
  }, {
    key: "renderStar",
    value: function renderStar() {
      var _this$props2 = this.props,
          isFilled = _this$props2.isFilled,
          isHalf = _this$props2.isHalf;

      if (isFilled) {
        return isHalf ? _react["default"].createElement(_starHalf["default"], null) : _react["default"].createElement(_starFill["default"], null);
      }

      return _react["default"].createElement(_starBordered["default"], null);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props3 = this.props,
          value = _this$props3.value,
          name = _this$props3.name,
          readOnly = _this$props3.readOnly;
      return _react["default"].createElement(_starContainer["default"], null, _react["default"].createElement(_starInput["default"], {
        type: "radio",
        id: this.starId,
        value: value,
        name: name,
        onChange: this.handleChange,
        disabled: readOnly
      }), _react["default"].createElement("label", {
        htmlFor: this.starId
      }, this.renderStar(), _react["default"].createElement(_AssistiveText["default"], {
        text: this.getAssitiveText()
      })));
    }
  }]);
  return Star;
}(_react.Component);

exports["default"] = Star;
Star.propTypes = {
  value: _propTypes["default"].number,
  onChange: _propTypes["default"].func,
  isFilled: _propTypes["default"].bool.isRequired,
  isHalf: _propTypes["default"].bool,
  name: _propTypes["default"].string,
  readOnly: _propTypes["default"].bool
};
Star.defaultProps = {
  value: 1,
  onChange: function onChange() {},
  name: undefined,
  readOnly: false,
  isHalf: false
};
