// react-rainbow-components/components/PicklistOption/pageObject/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/**
 * PicklistOption page object class.
 * @class
 */
function isPointWithinRect(point, rect) {
  var x = point.x,
      y = point.y;
  var left = rect.left,
      top = rect.top,
      right = rect.right,
      bottom = rect.bottom;
  return x >= left && y >= top && x <= right && y <= bottom;
}

var PagePicklistOption =
/*#__PURE__*/
function () {
  /**
   * Create a new PicklistOption page object.
   * @constructor
   * @param {string} rootElement - The selector of the PicklistOption root element.
   */
  function PagePicklistOption(rootElement, containerRect) {
    (0, _classCallCheck2["default"])(this, PagePicklistOption);
    this.rootElement = rootElement;
    this.containerRect = containerRect;
  }
  /**
   * Clicks the PicklistOption
   * @method
   */


  (0, _createClass2["default"])(PagePicklistOption, [{
    key: "click",
    value: function click() {
      this.rootElement.click();
    }
    /**
     * It moves the pointer over the PicklistOption
     * @method
     */

  }, {
    key: "hover",
    value: function hover() {
      var itemElement = this.rootElement.$('div[role="option"]');
      itemElement.moveTo();
    }
    /**
     * Get the label of the PicklistOption.
     * @method
     * @returns {string}
     */

  }, {
    key: "getLabel",
    value: function getLabel() {
      return this.rootElement.$('div[role="option"]').getText();
    }
    /**
     * Returns true when the PicklistOption is active.
     * @method
     * @returns {bool}
     */

  }, {
    key: "isActive",
    value: function isActive() {
      return this.rootElement.$('div[aria-selected="true"]').isExisting();
    }
    /**
     * Returns true when the PicklistOption is selected.
     * @method
     * @returns {bool}
     */

  }, {
    key: "isSelected",
    value: function isSelected() {
      return this.rootElement.$('li[data-selected="true"]').isExisting();
    }
    /**
     * Returns true when the PicklistOption is visible inside the menu container.
     * @method
     * @returns {bool}
     */

  }, {
    key: "isVisible",
    value: function isVisible() {
      var _this$rootElement$$$g = this.rootElement.$('div[role="option"]').getLocation(),
          x = _this$rootElement$$$g.x,
          y = _this$rootElement$$$g.y;

      var _this$rootElement$$$g2 = this.rootElement.$('div[role="option"]').getSize(),
          width = _this$rootElement$$$g2.width,
          height = _this$rootElement$$$g2.height;

      return this.rootElement.isDisplayedInViewport() && isPointWithinRect({
        x: x,
        y: y
      }, this.containerRect) && isPointWithinRect({
        x: x + width,
        y: y + height
      }, this.containerRect);
    }
    /**
     *  Wait until the option is visible.
     * @method
     */

  }, {
    key: "waitUntilIsVisible",
    value: function waitUntilIsVisible() {
      var _this = this;

      browser.waitUntil(function () {
        return _this.isVisible();
      });
    }
  }]);
  return PagePicklistOption;
}();

module.exports = PagePicklistOption;
