// react-rainbow-components/components/Input/pageObject/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/**
 * Input page object class.
 * @class
 */
var PageInput =
/*#__PURE__*/
function () {
  /**
   * Create a new Input page object.
   * @constructor
   * @param {string} rootElement - The selector of the Input root element.
   */
  function PageInput(rootElement) {
    (0, _classCallCheck2["default"])(this, PageInput);
    this.rootElement = rootElement;
  }
  /**
   * Clicks the input element.
   * @method
   */


  (0, _createClass2["default"])(PageInput, [{
    key: "click",
    value: function click() {
      $(this.rootElement).$('input').click();
    }
    /**
     * Clicks the label element.
     * @method
     */

  }, {
    key: "clickLabel",
    value: function clickLabel() {
      $(this.rootElement).$('label').click();
    }
    /**
     * Returns true when the input element has focus.
     * @method
     * @returns {bool}
     */

  }, {
    key: "hasFocusInput",
    value: function hasFocusInput() {
      return $(this.rootElement).$('input').isFocused();
    }
    /**
     * Type in the input element.
     * @method
     * @param {string} value - The value to type in the input element.
     */

  }, {
    key: "setValue",
    value: function setValue(value) {
      $(this.rootElement).$('input').setValue(value);
    }
    /**
     * Get the value typed in the input element.
     * @method
     * @returns {string}
     */

  }, {
    key: "getValue",
    value: function getValue() {
      return $(this.rootElement).$('input').getValue();
    }
  }]);
  return PageInput;
}();

module.exports = PageInput;
