// react-rainbow-components/components/ProgressCircular/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ProgressCircular;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _AssistiveText = _interopRequireDefault(require("./../AssistiveText"));

var _normalizeValue = _interopRequireDefault(require("./normalizeValue"));

var _progressRing = _interopRequireDefault(require("./progressRing"));

var _container = _interopRequireDefault(require("./styled/container"));

var _percentValue = _interopRequireDefault(require("./styled/percentValue"));

/**
 * ProgressCircular component communicates to the user the progress of a particular process.
 */
function ProgressCircular(props) {
  var value = props.value,
      variant = props.variant,
      assistiveText = props.assistiveText,
      className = props.className,
      style = props.style;
  var normalizedValue = (0, _normalizeValue["default"])(value);
  return _react["default"].createElement(_container["default"], {
    className: className,
    "aria-valuemin": "0",
    "aria-valuemax": "100",
    "aria-valuenow": normalizedValue,
    role: "progressbar",
    style: style
  }, _react["default"].createElement(_progressRing["default"], {
    variant: variant,
    percent: normalizedValue
  }), _react["default"].createElement(_percentValue["default"], {
    variant: variant
  }, "".concat(normalizedValue, "%")), _react["default"].createElement(_AssistiveText["default"], {
    text: assistiveText
  }));
}

ProgressCircular.propTypes = {
  /** The percentage value of the progress circular. It must be between 0 and 100. */
  value: _propTypes["default"].number,

  /** The variant of the progress circular. Valid values are brand and success. */
  variant: _propTypes["default"].oneOf(['brand', 'success', 'warning', 'error']),

  /** A description for assistive sreen readers. */
  assistiveText: _propTypes["default"].string,

  /** A CSS class for the outer element, in addition to the component's base classes. */
  className: _propTypes["default"].string,

  /** An object with custom style applied to the outer element. */
  style: _propTypes["default"].object
};
ProgressCircular.defaultProps = {
  value: 0,
  variant: 'brand',
  assistiveText: '',
  className: undefined,
  style: undefined
};
