// react-rainbow-components/components/AccordionSection/pageObject/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

/**
 * AccordoinSection page object class.
 * @class
 */
var PageAccordionSection =
/*#__PURE__*/
function () {
  /**
   * Create a new AccordionSection page object.
   * @constructor
   * @param {string} rootElement - The selector of the AccordoinSection root element.
   */
  function PageAccordionSection(rootElement) {
    (0, _classCallCheck2["default"])(this, PageAccordionSection);
    this.rootElement = rootElement;
  }
  /**
   * Clicks the button icon element.
   * @method
   */


  (0, _createClass2["default"])(PageAccordionSection, [{
    key: "clickButtonIcon",
    value: function clickButtonIcon() {
      $(this.rootElement).$('[data-id="accordion-section-summary"] > button').click();
    }
    /**
     * Returns true when the button icon has focus.
     * @method
     * @returns {bool}
     */

  }, {
    key: "hasFocusButtonIcon",
    value: function hasFocusButtonIcon() {
      return $(this.rootElement).$('[data-id="accordion-section-summary"] > button').isFocused();
    }
    /**
     * Returns true when the accordion section is expanded, false otherwise.
     * @method
     * @returns {bool}
     */

  }, {
    key: "isExpanded",
    value: function isExpanded() {
      return $(this.rootElement).$('[data-id="accordion-section-content"]').isDisplayed();
    }
    /**
     * Returns the label of the accordion section.
     * @method
     * @returns {string}
     */

  }, {
    key: "getLabel",
    value: function getLabel() {
      return $(this.rootElement).$('[title="Accordion Label"]').getText();
    }
  }]);
  return PageAccordionSection;
}();

module.exports = PageAccordionSection;
